contrib/swank-buffer-streams.lisp

```diff
@@ -0,0 +1,58 @@
+;;; swank-buffer-streams.lisp --- Streams that output to a buffer
+;;;
+;;; Authors: Ed Langley  <el-github@elangley.org>
+;;;
+;;; License: This code has been placed in the Public Domain.  All warranties
+;;;          are disclaimed.
+
+(in-package :swank)
+
+(defpackage :swank-buffer-streams
+  (:use :cl)
+  (:import-from :swank
+                defslimefun
+                add-hook
+                encode-message
+                send-event
+                find-thread
+                dcase
+                current-socket-io
+                send-to-emacs
+                current-thread-id
+                wait-for-event
+
+                *emacs-connection*
+                *event-hook*)
+  (:export initialize-buffer-stream-hooks
+           make-buffer-output-stream))
+
+(in-package :swank-buffer-streams)
+
+(defslimefun initialize-buffer-stream-hooks ()
+  (add-hook *event-hook* 'handle-events)
+  nil)
+
+(defun handle-events (_ event)
+  (declare (ignore _))
+  (dcase event
+    ((:make-stream-target &rest _) (declare (ignore _))
+     (encode-message event (current-socket-io))
+     t)
+    (((:stream-target-created) thread-id &rest _)
+     (declare (ignore _))
+     (send-event (find-thread thread-id) event)
+     t)
+    (t nil)))
+
+(defun get-temporary-identifier ()
+  (intern (symbol-name (gensym "BUFFER"))
+          :keyword))
```

Comment:
I think if someone is resetting *gensym-counter*, they're generally either trying to produce a symbol collision or they've restarted their repl.

I guess the second case might be a problem, but the first case is this working as expected.

As far as leaking keywords goes, the identifier is the second return value of MAKE-BUFFER-STREAMS: I guess we could use strings (or uninterned symbols, to keep them matchable by eql: but the emacs-side hashtable might be an EQL hash-table).  In my mind, at least, it's ok to leak keywords here because the user is requesting a much more memory intensive resource (emacs buffers) and the keyword is a handle for them to re-acquire that resource, should they lose the connection.

Reply:
Okey. They can always pass something else if they need to create a lot of these.